### ISSUE TITLE
PNG file bit depth is not recognized correctly

### ISSUE THREAD

**Issue by hverases** [SixLabors/ImageSharp#1954]
### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have verified that I am running the latest version of ImageSharp
- [x] I have verified if the problem exist in both `DEBUG` and `RELEASE` mode
- [x] I have searched [open](https://github.com/SixLabors/ImageSharp/issues) and [closed](https://github.com/SixLabors/ImageSharp/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description
When I load a png image the bit depth of the input file is always recognized as 32bpp, no matter if the input image actually has a bit depth of 8bpp. I think it would be a nice feature keeping the origin image bit depth between image conversions.

### Steps to Reproduce
Using an 8 bit depth png image as input:

![ModifiedImage1_8BitDepth](https://user-images.githubusercontent.com/10520121/150650225-80a7de4e-a621-4547-ba7d-d0fa9a11f414.png)

Running the following code outputs "32", both in debug and release:

```cs
var pngFile = "ModifiedImage1.png";

using (var image = Image.Load(pngFile))
{
    Console.WriteLine(image.PixelType.BitsPerPixel);
}
```
Debug data as extra info:

![ImageSharp_BitsPerPixel](https://user-images.githubusercontent.com/10520121/150650254-d2c4b522-05c8-42d7-833a-33ab63bd5239.png)

And I think that the expected value should be 8, right?

This is the repro image:

![ModifiedImage1](https://user-images.githubusercontent.com/10520121/150695579-191080ff-166a-4c01-9123-5f2c12dade2c.png)

It's from an old game running in 256 color mode.

But the issue also happens using 1 bit black and white images as input.

### System Configuration

- ImageSharp version: **1.0.4**
- Other ImageSharp packages and versions: **No other versions or packages used apart from SixLabors.ImageSharp 1.0.4**
- Environment (Operating system, version and so on): **Windows 10, Visual Studio 2022**
- .NET Framework version: **.NET 6**
- Additional information: **-**

**Comment by mikaellindemann** [comment]
Isn't this the depth of the in-memory representation of the (decoded) image?
It seems that after load, the pixel format is Rgba32 from your debugger.

**Comment by JimBobSquarePants** [comment]
@hverases why did you delete the issue template?

**Comment by hverases** [comment]
@mikaellindemann yeah, you are right, but what if I want to convert it to another format while keeping the original bit depth? Unless I enter manually the desired bit depth at the encoder, bit depth won't be the same as the original image. As an example, if I want to convert or process a batch of images with different bit depths the previous "trick" won't work.

@JimBobSquarePants sorry, I'm new posting issues on Github, and from what I have seen there are another issues that doesn't follow the template. Edit: sorry again, I haven't looked hard enough, more recent issues follow the template. Should I edit my first message following the template?

**Comment by antonfirsov** [comment]
> Should I edit my first message following the template?

Yes please. Also put enphasis on expected vs actual behavior, and attach repro image if possible.

(There are some issues not using the template but those are tracking internal work defined by the team.)

**Comment by hverases** [comment]
@JimBobSquarePants, @antonfirsov issue edited following the template.

**Comment by JimBobSquarePants** [comment]
Right.....

So, long story short we cannot automate that process for you. 

As @mikaellindemann  correctly pointed out that `PixelTypeInfo.BitsPerPixel` property represents the bit depth of the decoded image. v1.0.4 always loads that when using the non-generic version as `Rgba32` however V2 will optimize memory allocation by using the most appropriate pixel format to decode to based upon the detected color type and bit depth in the input stream https://github.com/SixLabors/ImageSharp/pull/1861 and you would see the number you expect.

However, and this is important, the information you require to determine the original bit depth is not contained within that property.  

In your screenshot you can see a `Metadata` property of type [`ImageMetadata`](https://docs.sixlabors.com/api/ImageSharp/SixLabors.ImageSharp.Metadata.ImageMetadata.html) that contains a method [`GetFormatMetadata<TFormatMetadata>(IImageFormat<TFormatMetadata>)`](https://docs.sixlabors.com/api/ImageSharp/SixLabors.ImageSharp.Metadata.ImageMetadata.html#SixLabors_ImageSharp_Metadata_ImageMetadata_GetFormatMetadata__1_SixLabors_ImageSharp_Formats_IImageFormat___0__) which can be used to return the decoded metadata (or an instance with sensible prepopulated values) from your image. For each format we actually provide some syntactic sugar via extension methods to that type to make it easier to do. For png this is [`GetPngMetadata()`](https://docs.sixlabors.com/api/ImageSharp/SixLabors.ImageSharp.MetadataExtensions.html#SixLabors_ImageSharp_MetadataExtensions_GetPngMetadata_SixLabors_ImageSharp_Metadata_ImageMetadata_). The returned [type](https://docs.sixlabors.com/api/ImageSharp/SixLabors.ImageSharp.Formats.Png.PngMetadata.html) in this instance contains everything you need to know about the input png. You can then use that information in your own code to determine the best encoding options for your changed format.

So why can't we automate this for you?

Well the answer is actually quite simple. In order to do so each individual encoder would then have to have intimate knowledge of all the available formats in order to do a conversion. Different image formats store bit depth via different metrics, for example png uses bits per sample or per palette index (not per pixel) whereas bmp uses per pixel. 

A long time ago we considered storing a central property on `ImageMetadata` to track this but since the bit depth and color type are so intrinsically tied with many formats it was seen as best (and much easier for us) to keep all relevant format specific information together.

**Comment by hverases** [comment]
First of all, thank you so much for your detailed reply, @JimBobSquarePants. I understand why you decided to not automate pixel format between different image formats.

Second, sorry for the late reply. I've been busy during the last days.

The thing is, I recently wrote a small console utility to modify raw pixel bytes from png files, and in order to do that, I needed to convert the image to bmp format first and then get the bytes of the raw, uncompressed pixel data. I made the conversion to bmp in memory, got the bytes from the stream, made the changes and then saved this modified array as a new png. I used Magick.NET for it, which I've been used a few years ago for another small command-line utilities. It worked perfectly, and it made the format conversions keeping the bit depth _automagically_, so I assumed that it would be the same using another library.

Magick.NET worked like a charm, but I didn't like the fact that it takes a lot of disk space and have different binaries for each platform. Then I found some info about ImageSharp, it tooks little space and was written using only dotnet managed code. It would be a bit slower, but execution time wasn't important for my use case.

Considering your answer, I have been experimenting with ImageSharp, first trying to get pixel bytes with the correct bit depth using an 8 bit png as input and later using a 1 bit png.

For 8 bit image, it was as simple as this:
```cs
var pngFile = "ModifiedImage1.png";

using (var image = Image.Load(pngFile))
{
    BmpEncoder encoder = new BmpEncoder();
    encoder.BitsPerPixel = (BmpBitsPerPixel)image.Metadata.GetPngMetadata().BitDepth.Value;
    image.SaveAsBmp("output.bmp", encoder);
}
```
This was to save the output bmp on disk to check that it has the same bit depth.

It can be done also getting the image info with the Identify method of the pertinent format decoder, PngDecoder in this case.

But for the 1 bit png this didn't work, and it surprised me because before opening this issue I have been looking for previous similar issues and I found #1623 PR, and it looks like it was successfully merged into master branch. The PR date was older than the release version I was using, and I confirmed later that this feature was not included in 1.0.4, Pixel1 value was not present in the BmpBitsPerPixel enum.

So I got 2.0.0-alpha.0.159 from MyGet and now it's working even with 1 bit depth images.

I think that I can close now this false issue because it wasn't really an issue but a misunderstanding of mine about the library inner workings. I will be glad if I can help someone with this message.

**Comment by JimBobSquarePants** [comment]
Now worries at all, It's very useful to have conversations documented like this for others and demonstrates the requirement for better documentation. 

>It would be a bit slower, but execution time wasn't important for my use case.

We're a LOT faster than Magick.NET for many operations so I'd be very curious to see where our sticky points are.

**Comment by hverases** [comment]
@JimBobSquarePants about ImageSharp being a bit slower than Magick.NET... I was totally wrong. It's the opposite.

When I was looking for a Magick.NET alternative I found [this article](https://devblogs.microsoft.com/dotnet/net-core-image-processing/) and it's not only outdated (2017), I guess I did an extra-quick reading of the article and misinterpreted the graphs... the numbers are execution times in milliseconds, so lower values are better.

It's also interesting how some functions are implemented taking advantage of SIMD instructions like SSE2 or AVX2, I didn't know something like this would be possible on dotnet managed code or C#. Anyways, this is out of my field, I work as a web developer doings CRUDs with more or less complexity based on design or architecture patterns, so I know little about algorithms optimization.
